lib/salus/scanners/osv/maven_osv.rb

```diff
@@ -0,0 +1,150 @@
+    end
+
+    def run
+      dependencies = find_dependencies
+      if dependencies.empty?
+        err_msg = "MavenOSV: Failed to parse any dependencies from the project."
+        report_stderr(err_msg)
+        report_error(err_msg)
+        return
+      end
+
+      @osv_vulnerabilities ||= fetch_vulnerabilities(MAVEN_OSV_ADVISORY_URL)
+      if @osv_vulnerabilities.nil?
+        err_msg = "No vulnerabilities found to compare."
+        bugsnag_notify("MavenOSV: #{err_msg}")
+        return report_error("MavenOSV: #{err_msg}")
+      end
+
+      # Fetch vulnerable dependencies.
+      # Dedupe and select Github Advisory over other sources if available.
+      results = []
+      grouped = match_vulnerable_dependencies(dependencies).group_by { |d| d[:ID] }
+      grouped.each do |_key, values|
+        vuln = {}
+        values.each do |v|
+          vuln = v if v[:Database] == GITHUB_DATABASE_STRING
+        end
+        results.append(vuln.empty? ? values[0] : vuln)
+      end
+
+      # Report scanner status
+      return report_success if results.empty?
+
+      report_failure
+      log(JSON.pretty_generate(results))
+    end
+
+    # Match if dependency version found is in the range of
+    # vulnerable dependency found.
+    def version_matching(version, version_ranges)
+      vulnerable_flag = false
+      version_found = SemVersion.new(version)
+      # If version range length is 1, then no fix available.
+      if version_ranges.length == 1
+        introduced = SemVersion.new(
+          version_ranges[0]["introduced"]
+        )
+        vulnerable_flag = true if version_found >= introduced
+      # If version range length is 2, then both introduced and fixed are available.
+      elsif version_ranges.length == 2
+        introduced = SemVersion.new(
+          version_ranges[0]["introduced"]
+        )
+        fixed = SemVersion.new(
+          version_ranges[1]["fixed"]
+        )
+        vulnerable_flag = true if version_found >= introduced && version_found < fixed
+      end
+
+      vulnerable_flag
+    end
+
+    # Compare vulnerabilities found with dependencies found
+    # and return vulnerable dependencies
+    def match_vulnerable_dependencies(dependencies)
+      results = []
+      dependencies.each do |dependency|
+        lib = "#{dependency['group_id']}:#{dependency['artifact_id']}"
+
+        unless dependency['version'].nil?
+          version = dependency['version']
+          # If version is of the format '${deps.version}', log it.
+          if version.match(/\${(.*)\.version}/)
+            bugsnag_notify("MavenOSV: Found #{lib}:#{version} with incompatible format.")
+            break
+          end
+
+          package_matches = @osv_vulnerabilities.select do |v|
+            v.dig("package", "name") == lib
+          end
+          package_matches.each do |m|
+            m["ranges"].each do |version_ranges|
+              if version_ranges["type"] == "SEMVER" || version_ranges["type"] == "ECOSYSTEM"
```

Comment:
should we be skipping all `version_ranges` that do not match the above?

Reply:
For now I have encountered 3 types - SEMVER, ECOSYSTEM, GIT.
So we ignore the GIT ones since it lists which git commit introduced the vulnerability. We don't have a way to compare git commits from the dependency list and also it gets covered under SEMVER, ECOSYSTEM.